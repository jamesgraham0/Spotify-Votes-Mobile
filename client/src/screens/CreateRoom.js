import { 
    StyleSheet, 
    Text, 
    View, 
    TouchableOpacity, 
    TextInput, 
    KeyboardAvoidingView, 
    Keyboard, 
    Linking,
    Alert
} from 'react-native';
import React, { useEffect, useState } from 'react';
import Ionicons from '@expo/vector-icons/Ionicons';
import uuid from 'react-native-uuid';
// import { createRoom } from '../reducers/reducer';
// import { useDispatch } from 'react-redux';
import service from '../utils/service';
import { socket } from '../utils/socket';
// import { Linking } from 'expo';
import * as Haptics from 'expo-haptics';

const CreateRoom = ({ navigation, route }) => {
    const { user } = route.params;
    const [roomName, setRoomName] = useState('');
    const [password, setPassword] = useState('');
    const url = 'https://open.spotify.com';
    
    const handleReturnToJoinOrCreateRoom = () => {
        navigation.navigate('JoinOrCreateRoom', { user: user });
    }
    
    const handleCreateRoom = async () => {
        Keyboard.dismiss();
        if (validNameAndPassword()) {
            const room = {
                name: roomName, 
                password: password,
                hostId: user.id,
                id: uuid.v4(),
                deviceId: await service.getDeviceId(),
                users: [user],
                currentlyPlaying: {},
                queue: [],
            }
            if (room.deviceId !== '' && room.deviceId !== undefined) {
                socket.emit('createRoom', room);
                navigation.navigate(
                    'Room', 
                    {room: room,
                    user: user}
                );
                await service.resetPlaybackToEmptyState();
            } else {
                const supported = await Linking.canOpenURL(url);
                if (supported) {
                    Alert.alert(
                        'Wait',
                        "Redirecting you to Spotify, toggle the play/pause button and then try creating a room again",
                        [
                            {
                                text: "Open Spotify",
                                onPress: async () => await Linking.openURL(url),
                            }
                        ]
                    );
                } else {
                    alert(`Open ${url} in your browser before creating a room`);
                }
            }
        } else {
            alert("Enter a room name and password to continue");
        }
    }
            
    const validNameAndPassword = () => {
        return roomName !== '' && password !== '';
    }
            
    return (
        <View style={styles.outerContainer}>
            <View style={styles.container}>
                <Text style={styles.createRoomText}>Create Room</Text>
            </View>
            <Text style={styles.instructionText}>Give your room a name and password. Others will use this password to join your room!</Text>
            <TouchableOpacity 
            onPress={() => {
                Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
                handleReturnToJoinOrCreateRoom();
            }} 
            style={styles.returnButton}
            >
                <Ionicons name="chevron-back-circle-outline" size={32} color="grey" />
            </TouchableOpacity>

            <KeyboardAvoidingView 
                behavior={Platform.OS === "ios" ? "padding" : "height"}
                style={styles.inputWrapper}
                >
                <TextInput 
                    style={styles.textInput} 
                    placeholder={'Project X 2.0...'} 
                    value={roomName} 
                    onChangeText={word => setRoomName(word)}
                    placeholderTextColor="#888" 
                    />
            
                <TextInput 
                    style={styles.textInput} 
                    placeholder={'Password123...'} 
                    value={password} 
                    onChangeText={word => setPassword(word)} 
                    placeholderTextColor="#888"
                />
                <TouchableOpacity onPress={() => {
                    Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Heavy);
                    handleCreateRoom()
                }}>
                    <View style={styles.addWrapper}>
                        <Text style={styles.addText}>+</Text>
                    </View>
                </TouchableOpacity>
            </KeyboardAvoidingView>
        </View>
    )
}

const styles = StyleSheet.create({
    outerContainer: {
        flex: 1,
        backgroundColor: '#191414',
        alignItems: 'center',
    },
    container: {
      flex: 1,
      backgroundColor: '#191414',
      alignItems: 'center',
    },
    createRoomText: {
        top: 100,
        color: '#fff',
        fontSize: 24,
    },
    returnButton: {
        position: 'absolute',
        top: 100,
        width: 50,
        height: 50,
        left: 30,
    },
    instructionText: {
        position: 'absolute',
        top: 150,
        left: 30,
        color: '#BBB',
        fontSize: 20,
        width: '70%',
        height: 90,
        margin: 10,
        textAlign: 'left',
    },
    inputWrapper: {
        backgroundColor: '#101010',
        position: 'absolute',
        top: 250,
        width: '90%',
        height: '30%',
        flexDirection: 'column',
        alignItems: 'center',
        borderRadius: 30,
        margin: 20,
        shadowColor: '#1DB954',
        shadowOpacity: 0.3,
        shadowOffset: { width: -2, height: -2 },
    },
    textInput: {
        margin: 30,
        padding: 10,
        width: 250,
        fontSize: 24,
        color: '#fff',
        backgroundColor: '#101010',
        borderStartColor: '#B026FF',
        borderStartWidth: 1,
        borderBottomColor: '#B026FF',
        borderBottomWidth: 2,
        borderBottomEndRadius: 100,
    },
    addWrapper: {
        margin: 70,
        width: 60,
        height: 60,
        borderRadius: 60,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 2,
        borderColor: '#1DB954',
        shadowOpacity: 1,
        shadowColor: '#1DB954',
        shadowOffset: { width: 0, height: 0 },
        shadowRadius: 5,
    },
    addText: {
        color: '#fff',
        fontSize: 24,
    }
  })
  

export default CreateRoom;