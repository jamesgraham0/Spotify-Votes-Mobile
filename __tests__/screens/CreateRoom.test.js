import React from 'react';
import renderer from 'react-test-renderer';
import CreateRoom from '../../client/src/screens/CreateRoom';

describe('CreateRoom', () => {
    it('should render the CreateRoom component', () => {
        const navigation = {};
        const route = { params: { user: {} } };
        const tree = renderer.create(<CreateRoom navigation={navigation} route={route} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});