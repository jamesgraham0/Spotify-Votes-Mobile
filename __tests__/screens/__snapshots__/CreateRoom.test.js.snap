// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRoom should render the CreateRoom component 1`] = `
<View
  style={
    {
      "backgroundColor": "#191414",
      "flex": 1,
      "height": 300,
    }
  }
>
  <View
    style={
      {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "borderRadius": 100,
            "elevation": 5,
            "height": 200,
            "opacity": 0.2,
            "position": "absolute",
            "right": 150,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 510,
            "width": 200,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 200,
            "bottom": -200,
            "elevation": 5,
            "height": 400,
            "opacity": 0.3,
            "position": "absolute",
            "right": -100,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "width": 400,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 50,
            "bottom": 280,
            "elevation": 5,
            "height": 100,
            "opacity": 0.1,
            "position": "absolute",
            "right": 90,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "width": 100,
          },
          {
            "backgroundColor": "#1DB954",
            "shadowColor": "#1DB954",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 100,
            "elevation": 5,
            "height": 200,
            "left": 30,
            "opacity": 0.4,
            "position": "absolute",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 140,
            "width": 200,
          },
          {
            "backgroundColor": "rgba(10, 10, 10, 1)",
            "shadowColor": "rgba(10, 10, 10, 1)",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 75,
            "elevation": 5,
            "height": 150,
            "opacity": 0.4,
            "position": "absolute",
            "right": 40,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 150,
            "width": 150,
          },
          {
            "backgroundColor": "rgba(10, 10, 10, 1)",
            "shadowColor": "rgba(10, 10, 10, 1)",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 25,
            "elevation": 5,
            "height": 50,
            "opacity": 0.2,
            "position": "absolute",
            "right": 150,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": 120,
            "width": 50,
          },
          {
            "backgroundColor": "#B026FF",
            "shadowColor": "#B026FF",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "borderRadius": 125,
            "elevation": 5,
            "height": 250,
            "opacity": 0.4,
            "position": "absolute",
            "right": -60,
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 6,
            "top": -60,
            "width": 250,
          },
          {
            "backgroundColor": "#B026FF",
            "shadowColor": "#B026FF",
          },
        ]
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(10, 10, 10, 0.5)",
        "flexDirection": "row",
        "height": "13%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 50,
          "left": 20,
          "opacity": 1,
          "position": "absolute",
          "top": "50%",
          "width": 50,
        }
      }
    >
      <Text />
    </View>
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 24,
          "maxWidth": "70%",
          "position": "absolute",
          "top": "50%",
        }
      }
    >
      Create Room
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#BBB",
        "fontSize": 22,
        "marginBottom": 10,
        "marginHorizontal": 60,
        "marginTop": 220,
      }
    }
  >
    Give your new room a name!
  </Text>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#101010",
        "borderRadius": 30,
        "flexDirection": "column",
        "height": "30%",
        "justifyContent": "space-around",
        "margin": 20,
        "shadowColor": "#1DB954",
        "shadowOffset": {
          "height": -2,
          "width": -2,
        },
        "shadowOpacity": 0.3,
        "width": "90%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Chill party..."
      placeholderTextColor="#888"
      style={
        {
          "backgroundColor": "#101010",
          "borderBottomColor": "#B026FF",
          "borderBottomEndRadius": 100,
          "borderBottomWidth": 2,
          "borderStartColor": "#B026FF",
          "borderStartWidth": 1,
          "color": "#fff",
          "fontSize": 24,
          "marginTop": 50,
          "padding": 10,
          "width": 250,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#191414",
          "borderColor": "#B026FF",
          "borderRadius": 50,
          "borderWidth": 2,
          "display": "flex",
          "height": 80,
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "#B026FF",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 5,
          "width": 300,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontSize": 24,
              "position": "absolute",
            },
            undefined,
          ]
        }
      >
        Create Room
      </Text>
    </View>
  </View>
</View>
`;
